cpp/src/buffer/pinned_memory_resource.cu

```diff
@@ -0,0 +1,119 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cuda_runtime_api.h>
+
+#include <cuda/experimental/memory_resource.cuh>
+
+#include <rapidsmpf/buffer/pinned_memory_resource.hpp>
+#include <rapidsmpf/error.hpp>
+
+namespace rapidsmpf {
+
+namespace {
+cuda::experimental::memory_pool_properties get_memory_pool_properties(
+    PinnedPoolProperties const& properties
+) {
+    return cuda::experimental::memory_pool_properties{
+        .initial_pool_size = properties.initial_pool_size,
+        .release_threshold = properties.release_threshold
+    };
+}
+}  // namespace
+
+// PinnedMemoryPool implementation
+struct PinnedMemoryPool::PinnedMemoryPoolImpl {
+    PinnedMemoryPoolImpl(
+        int numa_id, cuda::experimental::memory_pool_properties properties
+    )
+        : numa_id(numa_id), p_pool{numa_id, std::move(properties)} {}
+
+    int numa_id;
+    cuda::experimental::pinned_memory_pool p_pool;
+};
+
+PinnedMemoryPool::PinnedMemoryPool(int numa_id, PinnedPoolProperties properties)
+    : numa_id_(numa_id),
+      properties_(std::move(properties)),
+      impl_(
+          std::make_unique<PinnedMemoryPoolImpl>(
+              numa_id, get_memory_pool_properties(properties_)
+          )
+      ) {}
+
+PinnedMemoryPool::~PinnedMemoryPool() = default;
+
+// PinnedMemoryResource implementation
+struct PinnedMemoryResource::PinnedMemoryResourceImpl {
```

Comment:
why the `PinnedMemoryResourceImpl` ?

Reply:
This is very similar to what we’ve proposed as a way to make RMM ABI-stable.